### PR TITLE
Trend Micro enhancement

### PR DESCRIPTION
**Pull request by reutshal**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/19265
fixes: https://github.com/demisto/etc/issues/28346

## Description
Please only review the added implemented commands - I still need to refactor the old ones to the new format.
Also, the `usdo` name was changed to `udso` in the product.
Since this is a beta integration We changed the name in the integration as well.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation 

## Demisto Partner?
- [ ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/test_data/logs_list_command_mock.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            "SinceTime": 0
+        },
+        "NextPage": {
+            "PageToken": 5,
+            "SinceTime": 1592726567
+        },
+        "Next": null,
+        "Count": 5,
+        "Logs": [
+            "Jul 14 2020 00:49:42 myvxkp.manage.trendmicro.com CEF:0|Trend Micro|Apex Central|2019|WB:36|36|3|deviceExternalId=1 rt=Jun 21 2020 07:56:09 GMT+00:00 app=5 cnt=1 dpt=80 act=2 src=10.128.0.11 cs1Label=SLF_PolicyName cs1=Internal User Policy deviceDirection=2 cat=36 dvchost=CU-PRO1-8254-2 request=http://www.eicar.org/download/eicar.com.txt duser=TRENDMICROAPEX-\\admin shost=TRENDMICROAPEX- deviceProcessName=C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe cn3Label=Web_Reputation_Rating cn3=49 deviceFacility=Apex One cn2Label=SLF_SeverityLevel cn2=100 ",
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/test_data/logs_list_command_mock.json:17]: Check that its not a sensitive data

File: Packs/TrendMicroApex/ReleaseNotes/2_0_0.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+
+#### Integrations
+##### Trend Micro Apex (Beta)
+- The following commands were added:
+    - TODO complete RN
```
Comment (barchen1) [inline Packs/TrendMicroApex/ReleaseNotes/2_0_0.md:5]: Make sure to do this

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.yml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 category: Endpoint
 commonfields:
-  id: Trend Micro Apex
+  id: Trend Micro Apex_copy
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.yml:3]: Make sure to change it back to Apex

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.yml [hunk 4]
```diff
@@ -26,7 +26,7 @@ configuration:
 description: Trend Micro Apex central automation to manage agents and User-Defined
   Suspicious Objects
 display: Trend Micro Apex (Beta)
-name: Trend Micro Apex
+name: Trend Micro Apex_copy
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.yml:29]: same as here

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py [hunk 5]
```diff
@@ -202,6 +393,136 @@ def prodagent_restore_command(client: Client, args):
...
+        response_data.pop('Logs')
+        response = {
+            'Logs': parsed_logs_list,
+            'Data': response_data
+        }
+
+    return CommandResults(
+        readable_output=readable_output,
+        outputs_prefix='TrendMicroApex.Log',
+        outputs=response,
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py:420]: Add outputs to yaml file (make sure for all the commands)

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py [hunk 6]
```diff
@@ -115,47 +182,171 @@ def prodagent_restore(self, multi_match=False, entity_id="", ip_add="", mac_add=
...
 
-def test_module(client: Client, *_):
+def test_module(client: Client):
     """
     Performs basic get request to get item samples
     """
-    client.usdo_list()
-    return 'ok', None, None
+    client.udso_list()
+    return 'ok'
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py:318]: Make sure all the errors are presented correctly.

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py [hunk 7]
```diff
@@ -213,25 +534,51 @@ def main():
...
-        if command in commands:
-            return_outputs(*commands[command](client, demisto.args()))  # type: ignore
+        if command == 'test-module':
+            test_result = test_module(client)
+            return_results(test_result)
+
+        elif command == 'trendmicro-apex-udso-list':
+            return_outputs(*udso_list_command(client, demisto.args()))
+
+        elif command == 'trendmicro-apex-udso-list':
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py:552]: Its the same command as the one before

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py [hunk 8]
```diff
@@ -115,47 +182,171 @@ def prodagent_restore(self, multi_match=False, entity_id="", ip_add="", mac_add=
...
+
+        suffix = f'{self.suffix}/{log_type}{querystring}'
+        jwt_token = self.create_jwt_token(http_method='GET', api_path=suffix, headers='', request_body='')
+
+        headers = {
+            'Authorization': 'Bearer ' + jwt_token,
+            'Content-Type': 'application/json;charset=utf-8'
+        }
+
+        response = (self._http_request("GET", url_suffix=suffix, headers=headers))
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py:200]: why do you need ()

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py [hunk 9]
```diff
@@ -202,6 +393,136 @@ def prodagent_restore_command(client: Client, args):
...
+    response = client.udso_add_file(file_content_base64_string, file_name, file_scan_action, note)
+
+    if response.get('result_code') == 1:
+        readable_output = f'### The file "{file_name}" was added to the UDSO list successfully'
+
+    else:
+        readable_output = f'There has been a problem while trying to create the file: \n' \
+                          f'{response.get("result_description")}'
+
+    # TODO need to insert the file to the context?
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py:447]: File is already in context under entry

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py [hunk 10]
```diff
@@ -202,6 +393,136 @@ def prodagent_restore_command(client: Client, args):
...
+    if response:
+        human_readable_table = response.get('result_content')
+
+    readable_output = tableToMarkdown('Servers List', human_readable_table, headerTransform=string_to_table_header,
+                                      removeNull=True)
+    return CommandResults(
+        readable_output=readable_output,
+        outputs_prefix='TrendMicroApex.Server',
+        outputs=response,
+        outputs_key_field='',
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py:472]: why outputs_key_field empty ?

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py [hunk 11]
```diff
@@ -202,6 +393,136 @@ def prodagent_restore_command(client: Client, args):
...
+        human_readable_table = response.get('result_content')
+
+    readable_output = tableToMarkdown('Agents List', human_readable_table, headerTransform=string_to_table_header,
+                                      removeNull=True)
+
+    return CommandResults(
+        readable_output=readable_output,
+        outputs_prefix='TrendMicroApex.Agent',
+        outputs=response,
+        outputs_key_field='',
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py:496]: why outputs_key_field empty ?

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py [hunk 12]
```diff
@@ -202,6 +393,136 @@ def prodagent_restore_command(client: Client, args):
...
+
+    readable_output = tableToMarkdown('Security Agents with Endpoint Sensor enabled', human_readable_table,
+                                      headerTransform=string_to_table_header,
+                                      removeNull=True)
+
+    return CommandResults(
+        readable_output=readable_output,
+        outputs_prefix='TrendMicroApex.EndpointSensorSecurityAgent',
+        outputs=human_readable_table,
+        outputs_key_field='',
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py:521]: why outputs_key_field empty ?

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py [hunk 13]
```diff
@@ -115,47 +182,171 @@ def prodagent_restore(self, multi_match=False, entity_id="", ip_add="", mac_add=
...
+            'Authorization': 'Bearer ' + jwt_token,
+            'Content-Type': 'application/json;charset=utf-8'
+        }
+
+        response = (self._http_request("GET", url_suffix=suffix, headers=headers))
+
+        return response
+
+    @staticmethod
+    def create_filter_entry(entry_value, entry_type):
```
Comment (barchen1) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py:241]: please add an explanation for this func

**Review by barchen1 — DISMISSED** [review]
Please add unit tests for the helper functions
@Arsenikr Are we making it an official integration, if so please go over the rest of the commands.

**Comment by reutshal** [comment]
@ronykoz @barchen1 Please Ignore the commented `Fetch incident` related lines since I'm still waiting for answers,
As well as the investigation commands that still need to be implemented.

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex_test.py [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+
+
+MOCK_LOGS_LIST = load_test_data('./test_data/logs_list_command_mock.json')
+MOCK_ADD_FILE = load_test_data('./test_data/add_file_command_mock.json')
+MOCK_SERVERS_LIST = load_test_data('./test_data/servers_list_command_mock.json')
+MOCK_AGENTS_LIST = load_test_data('./test_data/agent_list_command_mock.json')
+MOCK_SENSORS_LIST = load_test_data('./test_data/sensors_list_command_mock.json')
+MOCK_HISTORICAL_INVESTIGATION = load_test_data('./test_data/historical_investigation_command_mock.json')
+MOCK_RESULT_LIST = load_test_data('./test_data/result_list_command_mock.json')
+MOCK_RESULT_LIST_BY_STATUS = load_test_data('./test_data/result_list_by_status_mock.json')
```
Comment (ronykoz) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex_test.py:32]: load these at the tests, you are using it only once

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py (hunk 6 above)
Comment (ronykoz) [reply Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py:882]: @reutshal any update here?
Let's add fetch params here as well

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py [hunk 15]
```diff
@@ -115,47 +265,665 @@ def prodagent_restore(self, multi_match=False, entity_id="", ip_add="", mac_add=
...
+            'Authorization': 'Bearer ' + self.create_jwt_token(http_method='PUT', api_path=self.suffix,
+                                                               headers='', request_body=json.dumps(request_data))}
+
+        response = self._http_request("PUT", self.suffix, headers=headers, data=json.dumps(request_data))
+        # return_error(response)
+        self.validate_response(response, error_message="Endpoint sensors list operation failed")
+        return response
+
+    @staticmethod
+    def parse_cef_logs_to_dict_logs(response):
```
Comment (ronykoz) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.py:447]: this is a util function, not sure that the client is the right place for that

**Comment by ronykoz** [comment]
@reutshal let's fix the comments here and those we discussed in person and merge this PR.

**Comment by reutshal** [comment]
@kirbles19 Can you please go over the README file: https://github.com/demisto/content/blob/TrendMicroApex-integration/Packs/TrendMicroApex/Integrations/TrendMicroApex/README.md
And the yml: https://github.com/demisto/content/blob/TrendMicroApex-integration/Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.yml
Thanks!

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/README.md [hunk 16]
```diff
@@ -0,0 +1,759 @@
+Trend Micro Apex central automation to manage agents and User-Defined Suspicious Objects
+This integration was integrated and tested with version 1.0 of Trend Micro Apex_copy
```
Comment (ronykoz) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/README.md:2]:
```suggestion
This integration was integrated and tested with version 1.0 of Trend Micro Apex
```

Make sure there are no copies

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/README.md [hunk 17]
```diff
@@ -0,0 +1,759 @@
...
+## Configure Trend Micro Apex_copy on Cortex XSOAR
```
Comment (ronykoz) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/README.md:3]: same here and everywhere else

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.yml [hunk 18]
```diff
@@ -23,9 +23,14 @@ configuration:
   name: proxy
   required: false
   type: 8
+- display: Fetch incidents
```
Comment (ronykoz) [inline Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.yml:26]: didn't we say we are removing this until they respond?

**Review by ronykoz — COMMENTED** [review]
let's mention the BC break in the RN
Did you remove all the needed commands?

**Comment by reutshal** [comment]
> let's mention the BC break in the RN
> Did you remove all the needed commands?

Actually we didn't break BC, I only deprecated the old commands.
And yes, only the relevant commands are here

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/README.md (hunk 16 above)
Comment (reutshal) [reply Packs/TrendMicroApex/Integrations/TrendMicroApex/README.md:2]: Fixing

File: Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.yml (hunk 18 above)
Comment (reutshal) [reply Packs/TrendMicroApex/Integrations/TrendMicroApex/TrendMicroApex.yml:26]: Yes, removing

**Comment by ronykoz** [comment]
> > let's mention the BC break in the RN
> > Did you remove all the needed commands?
> 
> Actually we didn't break BC, I only deprecated the old commands.
> And yes, only the relevant commands are here

you should add the old commands to the python file than, because the usdo is not there.

**Review by ronykoz — APPROVED** [review]

**Comment by ronykoz** [comment]
@reutshal seems that you have a validation failure

**Comment by reutshal** [comment]
> @reutshal seems that you have a validation failure
That's an sdk bug that should be resolved today with the release

**Comment by kirbles19** [comment]
@reutshal YML is done also.

**Pull request merged**